realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java

```diff
@@ -361,6 +384,17 @@ private void emitValidateTableMethod(JavaWriter writer) throws IOException {
                         fieldTypeSimpleName, fieldName);
                 writer.endControlFlow();
 
+                // make sure that nullability matches
+                if (metadata.isNullable(field)) {
+                    writer.beginControlFlow("if (!table.isColumnNullable(%s))", staticFieldIndexVarName(field));
+                    writer.emitStatement("throw new RealmMigrationNeededException(transaction.getPath(), \"Add annotation @Required or @PrimaryKey to field '%s'\")", fieldName);
+                    writer.endControlFlow();
+                } else {
+                    writer.beginControlFlow("if (table.isColumnNullable(%s))", staticFieldIndexVarName(field));
+                    writer.emitStatement("throw new RealmMigrationNeededException(transaction.getPath(), \"Remove annotation @Required or @PrimaryKey from field '%s'\")", fieldName);
```

Comment:
Perhaps rewrite to "Field foo is nullable when it should not be. <Solution to problem>"